### PR TITLE
[addons] Bring back audioengine addon interface (not ADSP)

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description
This is a mandatory for Webbrowser addon system!

There are no changes on Audio Engine and from usage mostly equal to other
audioengine calls in Kodi.

All addons not need a recompile with them, on `AddonToKodiFuncTable_Addon` the `AddonToKodiFuncTable_kodi_audioengine* kodi_audioengine;` was still defined.

## How Has This Been Tested?
Tested on Webbrowser addon system

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/interfaces/AudioEngine.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+using namespace kodi; // addon-dev-kit namespace
+using namespace kodi::audioengine; // addon-dev-kit namespace
+
+namespace ADDON
+{
+
+void Interface_AudioEngine::Init(AddonGlobalInterface* addonInterface)
+{
+  addonInterface->toKodi->kodi_audioengine = (AddonToKodiFuncTable_kodi_audioengine*)malloc(sizeof(AddonToKodiFuncTable_kodi_audioengine));
```
Comment (thebrid) [inline xbmc/addons/interfaces/AudioEngine.cpp]: Why `malloc` rather than `new`?

Comment (AlwinEsch) [reply xbmc/addons/interfaces/AudioEngine.cpp]: The generated array is for a `C` code table, I'm not sure that a `new` outlets with pure "C" code might not generate traceable errors. Think to 98% the `new` works also always but there are 2% left ;-).

But good for the note, have forgot to change them to `static_cast<AddonToKodiFuncTable_kodi_audioengine*>(...)`.

File: xbmc/addons/interfaces/AudioEngine.cpp [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+  addonInterface->toKodi->kodi_audioengine->aestream_get_sample_rate = aestream_get_sample_rate;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_data_format = aestream_get_data_format;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_resample_ratio = aestream_get_resample_ratio;
+  addonInterface->toKodi->kodi_audioengine->aestream_set_resample_ratio = aestream_set_resample_ratio;
+}
+
+void Interface_AudioEngine::DeInit(AddonGlobalInterface* addonInterface)
+{
+  if (addonInterface->toKodi && /* <-- Safe check, needed so long old addon way is present */
+      addonInterface->toKodi->kodi_audioengine)
```
Comment (thebrid) [inline xbmc/addons/interfaces/AudioEngine.cpp]: The second check here is redundant.

Comment (AlwinEsch) [reply xbmc/addons/interfaces/AudioEngine.cpp]: This has created here on Linux crashes, where on PVR and Game addons (still in old interfaces) and `addonInterface->toKodi` not becomes set (from here https://github.com/xbmc/xbmc/blob/master/xbmc/addons/binary-addons/AddonDll.cpp#L184).

On new becomes this used: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/binary-addons/AddonDll.cpp#L264

Comment (AlwinEsch) [reply xbmc/addons/interfaces/AudioEngine.cpp]: Maybe I understand wrong and you mean to leave `addonInterface->toKodi->kodi_audioengine` away? And let the check to `free` itself?

Comment (thebrid) [reply xbmc/addons/interfaces/AudioEngine.cpp]: @AlwinEsch yes, the latter. `free()` (or `delete`) will do nothing on a null pointer. `clang-tidy` [complains](https://clang.llvm.org/extra/clang-tidy/checks/readability-delete-null-pointer.html) if we check this. :smile:

Comment (AlwinEsch) [reply xbmc/addons/interfaces/AudioEngine.cpp]: Have removed the `addonInterface->toKodi->kodi_audioengine` check

**Pull request merged**